source/io/Stream.cpp

```diff
@@ -13,132 +13,172 @@
+            aws_input_stream_vtable InputStream::s_vtable = {
+                InputStream::s_Seek,
+                InputStream::s_Read,
+                InputStream::s_GetStatus,
+                InputStream::s_GetLength,
+                InputStream::s_Destroy,
+            };
 
-    return AWS_OP_SUCCESS;
-}
+            InputStream::InputStream(Aws::Crt::Allocator *allocator)
+            {
+                m_allocator = allocator;
+                m_underlying_stream =
+                    static_cast<aws_input_stream *>(aws_mem_calloc(m_allocator, 1, sizeof(aws_input_stream)));
 
-static void s_aws_input_stream_cpp_destroy(struct aws_input_stream *stream)
-{
-    aws_input_stream_cpp_impl *impl = static_cast<aws_input_stream_cpp_impl *>(stream->impl);
-    impl->stream = nullptr;
-    aws_mem_release(stream->allocator, stream);
-}
+                if (!m_underlying_stream)
+                {
+                    m_good = false;
+                    return;
+                }
 
-static struct aws_input_stream_vtable s_aws_input_stream_cpp_vtable = {s_aws_input_stream_cpp_seek,
-                                                                       s_aws_input_stream_cpp_read,
-                                                                       s_aws_input_stream_cpp_get_status,
-                                                                       s_aws_input_stream_cpp_get_length,
-                                                                       s_aws_input_stream_cpp_destroy};
+                m_good = true;
+                m_underlying_stream->impl = this;
+                m_underlying_stream->allocator = m_allocator;
+                m_underlying_stream->vtable = &s_vtable;
+            }
 
-namespace Aws
-{
-    namespace Crt
-    {
-        namespace Io
-        {
-            struct aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
+            StdIOStreamInputStream::StdIOStreamInputStream(
+                std::shared_ptr<Aws::Crt::Io::IStream> stream,
                 Aws::Crt::Allocator *allocator) noexcept
+                : InputStream(allocator), m_stream(std::move(stream))
             {
-                struct aws_input_stream *input_stream = NULL;
-                struct aws_input_stream_cpp_impl *impl = NULL;
-
-                aws_mem_acquire_many(
-                    allocator,
-                    2,
-                    &input_stream,
-                    sizeof(struct aws_input_stream),
-                    &impl,
-                    sizeof(struct aws_input_stream_cpp_impl));
-
-                if (!input_stream)
+            }
+
+            bool StdIOStreamInputStream::Read(ByteBuf &buffer) noexcept
+            {
+                m_stream->read(reinterpret_cast<char *>(buffer.buffer + buffer.len), buffer.capacity - buffer.len);
+                buffer.len += static_cast<size_t>(m_stream->gcount());
+
+                return true;
```

Comment:
should be be checking `good()`/`bool()` after reading? I genuinely don't know

Reply:
I've been debating this one, I think we should just rely on GetStatus(), as std::io_stream only returns 0 even on a failure.

Reply:
The default implementations in aws-c-io reserve returning an error, to something that actually goes wrong, but not for end of stream.

Reply:
actually turns out I was wrong about this, fixed.